### PR TITLE
activerecord guide: clarified roles of valid method and errors object

### PR DESCRIPTION
**Pull request by benjiwheeler**
Per issue https://github.com/rails/rails/issues/20792

### CODE REVIEW

**Comment by trosborn** [comment]
Thanks for the PR. Could you add the [ci-skip] tag to your commit message?

File: guides/source/active_record_validations.md [hunk 1]
```diff
@@ -991,7 +1000,8 @@ class Invoice < ActiveRecord::Base
 end
 ```
 
-By default such validations will run every time you call `valid?`. It is also
+By default such validations will run every time `valid?` is called, whether
+it is called by your code or by Rails before the object is saved. It is also
```
Comment (matthewd) [inline guides/source/active_record_validations.md]: Hmm... by this point, it feels like we're over-selling the "Rails will call `valid?` while saving" bit -- which (AFAIK) isn't even part of any published contract. (We promise that AR will validate on save, and that `valid?` will validate... but not necessarily that we'll use that particular method internally.)

**Comment by benjiwheeler** [comment]
I see your point. How about:

"By default, validations will run every time you call `valid?` or save the object. But it is possible to more finely control when to run these custom validations by..."

**Comment by benjiwheeler** [comment]
Updated per comments from @trosborn and @matthewd

**Comment by benjiwheeler** [comment]
Please let me know if the changes I made per your comments look good.

File: guides/source/active_record_validations.md [hunk 2]
```diff
@@ -149,8 +149,10 @@ false` as an argument. This technique should be used with caution.
 
 ### `valid?` and `invalid?`
 
-To verify whether or not an object is valid, Rails uses the `valid?` method.
-You can also use this method on your own. `valid?` triggers your validations
+Before saving an ActiveRecord object, Rails runs your validations.
+If these validations produce any errors, Rails does not save the object.
```
Comment (arthurnn) [inline guides/source/active_record_validations.md]: dont get it why do we need those 2 lines here. Do we need to setup this context?

**Comment by arthurnn** [comment]
Not sure if this is adding much value TBH, But I am new to this. @senny might have more context in this.

File: guides/source/active_record_validations.md [hunk 3]
```diff
@@ -972,6 +975,10 @@ class method, passing in the symbols for the validation methods' names.
 You can pass more than one symbol for each class method and the respective
 validations will be run in the same order as they were registered.
 
+The `valid?` method will verify that the errors collection is empty,
+so your custom validation methods should add errors to it when you
+wish validation to fail:
```
Comment (senny) [inline guides/source/active_record_validations.md]: Doesn't the very first sentence in this section cover that already?

> You can also create methods that verify the state of your models and add messages to the errors collection when they are invalid.

File: guides/source/active_record_validations.md [hunk 4]
```diff
@@ -991,9 +998,10 @@ class Invoice < ActiveRecord::Base
 end
 ```
 
-By default such validations will run every time you call `valid?`. It is also
-possible to control when to run these custom validations by giving an `:on`
-option to the `validate` method, with either: `:create` or `:update`.
+By default, such validations will run every time you call `valid?`
+or save the object. But it is also possible to control when to run these
+custom validations by giving an `:on` option to the `validate` method,
+with either: `:create` or `:update`.
```
Comment (senny) [inline guides/source/active_record_validations.md]: I don't like this change. It should be clear by now, that `save` and `create` call `valid?` before actually saving the record. This is just duplicating that fact. The essence here is `valid?` and that the reader understand what methods will call `valid?`.

**Comment by senny** [comment]
I kind of agree with @arthurnn. This adds a lot of duplicated information. We should establish the fact that validations are only run when `valid?` is called. And what methods depend on valid before doing their thing (eg. `save`, `create` etc.)

**Comment by benjiwheeler** [comment]
At least for this relatively new rails developer, this doc wasn't clear enough previously that `valid?` was run not just "every time you call" it in code you add but also when rails calls it, which of course in a convention-over-configuration paradigm is especially blurred; and not clear enough previously that `valid?` failed iff the errors array was non-empty. i knew that I "can" add messages when errors occur, but not that validity depended entirely on whether or not i added messages.

**Comment by senny** [comment]
I understand. Let's finde a way to make the guide more clear.

@benjiwheeler There is a section ["When Does Validation Happen?"](https://github.com/benjiwheeler/rails/blob/validation_guide_ret_val/guides/source/active_record_validations.md#when-does-validation-happen) early in the guide. What do you feel is missing there that we need to reiterate in `valid?` that `save` and `create` call validations but `new` doesn't?

/cc @zzak

**Comment by benjiwheeler** [comment]
it's not so much that i think we need to reiterate that, but rather that what we do say should be clear:
1) clear about when `valid?` is called: "every time you call" `valid?` is not a good shorthand for this.
2) clear about the fact that `valid?` fails iff the errors array is non-empty. saying you "can" add messages when errors occur doesn't communicate this clearly enough.

i think there's a bit of a philosophical divide in documentation writing, with validity to each side... some think documentation should be like code, with behavior derived from the text and DRY. others, like me, feel that small portions of text should not assume knowledge of everything else, and that this requires some redundancy and repetition. 

i think if it's reasonably possible to understand something wrong -- not just incompletely -- from reading any portion of the text, that's a problem.

**Comment by sgrif** [comment]
I think the duplication is fine here, and adds clarity for newer developers who are reading this. The behavior of validations has been a sticking point with a lot of my students, and I think this would help.

**Comment by sgrif** [comment]
Please add `[ci skip]` to the commit message for documentation changes in the future

**Comment by benjiwheeler** [comment]
Thanks, will do!

**Pull request merged**